File: windows/Hinge.Core/ProtocolCompression.cs
using System.Buffers.Binary;
using System.IO.Compression;

namespace Hinge.Core;

/// <summary>
/// Optional compression for large JSON/control payloads.
///
/// The envelope is deliberately a new message type instead of a header flag so
/// that protocol v1 peers can continue to parse ordinary frames unchanged.
/// Compression is only enabled after the peer advertises the same capability
/// during the session identity exchange.
/// </summary>
public static class ProtocolCompression
{
    public const string Capability = "control-compression-zlib-v1";
    public const byte EnvelopeVersion = 1;
    public const int EnvelopeHeaderSize = 8;
    private const int MinimumPayloadSize = 2048;

    private static readonly MessageType[] CompressibleTypes =
    {
        MessageType.TextMessage,
        MessageType.SyncManifestRequest,
        MessageType.SyncManifestResponse,
        MessageType.SyncPullRequest,
        MessageType.ClipboardEvent,
        MessageType.NotificationEvent,
        MessageType.NotificationAction,
        MessageType.ToolCommand,
        MessageType.ToolResult
    };

    public static bool IsCompressible(MessageType type, int payloadLength) =>
        payloadLength >= MinimumPayloadSize &&
        Array.IndexOf(CompressibleTypes, type) >= 0;

    public static bool TryCompress(
        MessageType type,
        byte[] payload,
        out byte[] compressedPayload)
    {
        compressedPayload = Array.Empty<byte>();
        if (!IsCompressible(type, payload.Length) ||
            payload.Length > ProtocolFrame.MaxPayloadSize)
        {
            return false;
        }

        using var output = new MemoryStream();
        try
        {
            using (var zlib = new ZLibStream(output, CompressionLevel.Fastest, leaveOpen: true))
            {
                zlib.Write(payload, 0, payload.Length);
            }
        }
        catch (Exception)
        {
            // Compression is an optimization. A runtime/provider failure must
            // never prevent the uncompressed control frame from being sent.
            return false;
        }

        byte[] compressed = output.ToArray();
        if (compressed.Length + EnvelopeHeaderSize >= payload.Length)
        {
            return false;
        }

        compressedPayload = new byte[EnvelopeHeaderSize + compressed.Length];
        compressedPayload[0] = EnvelopeVersion;
        compressedPayload[1] = 0; // Reserved flags; must remain zero for v1.
        BinaryPrimitives.WriteUInt16BigEndian(
            compressedPayload.AsSpan(2, 2),
            (ushort)type);
        BinaryPrimitives.WriteUInt32BigEndian(
            compressedPayload.AsSpan(4, 4),
            (uint)payload.Length);
        compressed.CopyTo(compressedPayload.AsSpan(EnvelopeHeaderSize));
        return true;
    }

    public static bool TryDecompress(
        byte[] envelope,
        out MessageType type,
        out byte[] payload)
    {
        type = MessageType.ToolResult;
        payload = Array.Empty<byte>();

        if (envelope.Length <= EnvelopeHeaderSize ||
            envelope[0] != EnvelopeVersion ||
            envelope[1] != 0)
        {
            return false;
        }

        type = (MessageType)BinaryPrimitives.ReadUInt16BigEndian(
            envelope.AsSpan(2, 2));
        uint expectedLength = BinaryPrimitives.ReadUInt32BigEndian(
            envelope.AsSpan(4, 4));
        if (!IsCompressible(type, (int)Math.Min(expectedLength, int.MaxValue)) ||
            expectedLength > ProtocolFrame.MaxPayloadSize)
        {
            return false;
        }

        try
        {
            using var input = new MemoryStream(
                envelope,
                EnvelopeHeaderSize,
                envelope.Length - EnvelopeHeaderSize,
                writable: false);
            using var zlib = new ZLibStream(input, CompressionMode.Decompress);
            using var output = new MemoryStream((int)expectedLength);
            byte[] buffer = new byte[64 * 1024];
            int read;
            while ((read = zlib.Read(buffer, 0, buffer.Length)) > 0)
            {
                if (output.Length + read > expectedLength)
                {
                    return false;
                }
                output.Write(buffer, 0, read);
            }

            if (output.Length != expectedLength)
            {
                return false;
            }

            payload = output.ToArray();
            return true;
        }
        catch (InvalidDataException)
        {
            return false;
        }
        catch (IOException)
        {
            return false;
        }
    }
}

File: android/lib/core/protocol_compression.dart
import 'dart:io';
import 'dart:typed_data';

import 'protocol_frame.dart';

/// Optional compression for large JSON/control payloads.
///
/// The envelope matches windows/Hinge.Core/ProtocolCompression.cs:
/// version(1) + flags(1) + inner message type(2, big-endian) +
/// uncompressed length(4, big-endian) + zlib payload.
class ProtocolCompression {
  static const String capability = 'control-compression-zlib-v1';
  static const int envelopeVersion = 1;
  static const int envelopeHeaderSize = 8;
  static const int _minimumPayloadSize = 2048;

  static const Set<MessageType> _compressibleTypes = {
    MessageType.textMessage,
    MessageType.syncManifestRequest,
    MessageType.syncManifestResponse,
    MessageType.syncPullRequest,
    MessageType.clipboardEvent,
    MessageType.notificationEvent,
    MessageType.notificationAction,
    MessageType.toolCommand,
    MessageType.toolResult,
  };

  static bool isCompressible(MessageType type, int payloadLength) =>
      payloadLength >= _minimumPayloadSize && _compressibleTypes.contains(type);

  static Uint8List? tryCompress(MessageType type, Uint8List payload) {
    if (!isCompressible(type, payload.length) ||
        payload.length > ProtocolFrame.maxPayloadSize) {
      return null;
    }

    try {
      final compressed = Uint8List.fromList(
        ZLibCodec(level: 1).encode(payload),
      );
      if (compressed.length + envelopeHeaderSize >= payload.length) {
        return null;
      }

      final envelope = Uint8List(envelopeHeaderSize + compressed.length);
      final data = ByteData.sublistView(envelope);
      envelope[0] = envelopeVersion;
      envelope[1] = 0;
      data.setUint16(2, type.value, Endian.big);
      data.setUint32(4, payload.length, Endian.big);
      envelope.setRange(envelopeHeaderSize, envelope.length, compressed);
      return envelope;
    } on Object {
      // Compression is an optimization. A provider/runtime failure must not
      // interrupt the ordinary uncompressed protocol path.
      return null;
    }
  }

  static CompressedControlPayload? tryDecompress(Uint8List envelope) {
    if (envelope.length <= envelopeHeaderSize ||
        envelope[0] != envelopeVersion ||
        envelope[1] != 0) {
      return null;
    }

    final data = ByteData.sublistView(envelope);
    final type = MessageType.fromValue(data.getUint16(2, Endian.big));
    final expectedLength = data.getUint32(4, Endian.big);
    if (!isCompressible(type, expectedLength) ||
        expectedLength > ProtocolFrame.maxPayloadSize) {
      return null;
    }

    try {
      final decoded = Uint8List.fromList(
        ZLibCodec().decode(envelope.sublist(envelopeHeaderSize)),
      );
      if (decoded.length != expectedLength) return null;
      return CompressedControlPayload(type: type, payload: decoded);
    } on Object {
      return null;
    }
  }
}

class CompressedControlPayload {
  final MessageType type;
  final Uint8List payload;

  const CompressedControlPayload({required this.type, required this.payload});
}
